File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/Ack.java
package net.nemerosa.ontrack.model;

import lombok.Data;

import java.util.Optional;

@Data
public class Ack {

    public static final Ack OK = new Ack(true);
    public static final Ack NOK = new Ack(false);

    public static Ack validate(boolean test) {
        return test ? OK : NOK;
    }

    public static Ack validate(Optional<?> optional) {
        return optional != null ? validate(optional.isPresent()) : NOK;
    }

    public static Ack one(int count) {
        return validate(count == 1);
    }

    private final boolean success;

    public Ack and(Ack ack) {
        return validate(isSuccess() && ack.isSuccess());
    }

    public Ack or(Ack ack) {
        return validate(isSuccess() || ack.isSuccess());
    }

}

File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/Ack.kt
package net.nemerosa.ontrack.model

import java.util.*

data class Ack(
    val success: Boolean,
) {

    fun and(ack: Ack): Ack {
        return validate(success && ack.success)
    }

    fun or(ack: Ack): Ack {
        return validate(success || ack.success)
    }

    companion object {
        @JvmField
        val OK: Ack = Ack(true)
        @JvmField
        val NOK: Ack = Ack(false)

        fun validate(test: Boolean): Ack {
            return if (test) OK else NOK
        }

        fun validate(optional: Optional<*>?): Ack {
            return if (optional != null) validate(optional.isPresent) else NOK
        }

        @JvmStatic
        fun one(count: Int): Ack {
            return validate(count == 1)
        }
    }
}
